intrahost.py

```diff
@@ -129,157 +169,211 @@ def parser_tabfile_rename(parser=argparse.ArgumentParser()):
-
-            # define genotypes and fractions
-            iSNVs = {}
-            rows = dict(rows)
-            for s in samples:
-                if s in rows:
-                    consAllele = consAlleles[s]
-                    # we have iSNV data on this sample
-                    tot_n = sum(n for a,n in rows[s])
-                    iSNVs[s] = {}
-                    for a,n in rows[s]:
-                        f = float(n)/tot_n
-                        if a.startswith('I'):
-                            # insertion allele is first ref base, plus inserted bases, plus subsequent ref bases
-                            a = consAllele[0] + a[1:] + consAllele[1:]
-                        elif a.startswith('D'):
-                            # deletion is the first ref base, plus remaining ref seq with the first few positions dropped off
-                            a = consAllele[0] + consAllele[1+int(a[1:]):]
-                        elif a in ('i','d'):
-                            # this is vphaser's way of saying the "reference" (majority/consensus) allele, in the face of other indel variants
-                            a = consAllele
+        # one reference chrom at a time
+        with open(refFasta, 'rU') as inf:
+            for ref_sequence in Bio.SeqIO.parse(inf, 'fasta'):
+
+                # read in all iSNVs for this chrom and map to reference coords
+                data = []
+                for s in samples:
+                    for row in util.file.read_tabfile(samp_to_isnv[s]):
+                        s_chrom = samp_to_cmap[s].mapBtoA(ref_sequence.id)
+                        if row[0] == s_chrom:
+                            row = {'sample':s, 'CHROM':ref_sequence.id,
+                                's_chrom':s_chrom, 's_pos':int(row[1]),
+                                's_alt':row[2], 's_ref': row[3],
+                                'n_libs':row[7], 'lib_bias': row[8],
+                                'alleles':list(x.split(':') for x in row[9:] if x)}
+                            # make a sorted allele list
+                            row['allele_counts'] = list(sorted(
+                                [(a,int(f)+int(r)) for a,f,r in row['alleles']],
+                                key=(lambda x:x[1]), reverse=True))
+                            # reposition vphaser deletions minus one to be consistent with
+                            # VCF conventions
+                            if row['s_alt'].startswith('D'):
+                                for a,n in row['allele_counts']:
+                                    assert a[0] in ('D','i')
+                                row['s_pos'] = row['s_pos']-1
+                            # map position back to reference coordinates
+                            row['POS'] = samp_to_cmap[s].mapAtoB(s_chrom, row['s_pos'], side = -1)[1]
+                            row['END'] = samp_to_cmap[s].mapAtoB(s_chrom, row['s_pos'], side = 1)[1]
```

Comment:
This will raise if position is outside either end of the reference sequence. Can we safely assume that can't happen? If not, do we want to throw out any iSNVs beyond the reference ends? Similarly, down below, when we mapBtoA, can we assume that consensus is full length of reference? If not, we can treat it as if it were padded with Ns. Unless we assure in advance that reference and all consensuses have the same extent, I am inclined to change CoordMapper to return None rather than raising when past end so the caller can check result instead of putting a try block; generally, I save raise for cases where the caller has done something wrong, not for cases where the caller has no way to know in advance if the call will be valid or not.